### PR TITLE
common: optimize objects memory allocations using pools

### PR DESCRIPTION
**Pull request by orozery**
A significant time is spent in allocating memory for class objects on heap.
i.e., in calls to c++ operator new(long) and operator delete(void*).
This patch creates by-size pools to be used for allocating common objects,
which are relatively large, such as MOSDOp, Op, and librbd disaptch specs.

Implemented a meta class called PooledObject, which implements operator new and operator delete.
Then, simply make class such as MOSDOp inherit from PooledObject, and then every call to `new MOSDOp` will use the allocation implemented in PooledObject.

The classes I applied this to are upto 1K of size (most of them are just around 200 bytes).
So for example, a burst of 10000 MOSDOp's will add up to ~10MB of memory.

Also implemented Pooled Allocator for allocating memory on STL containers, currently applied only to the std::map of requests in SimpleSchedulerObjectDispatch.

See comparison of performance between the glibc operator new, and the pooled operator new:
image

### CODE REVIEW

**Comment by rzarzynski** [comment]
It seems the perf comparison comes from a `librbd` client. Which memory allocator has been used? IIRC we don't `tcmalloc` everywhere; clients / tests go with the vanilla one from glibc. There might be a signification difference in allocation costs, so let's be careful about benchmarking.

Anyway, this looks promising!

**Comment by markhpc** [comment]
Oh, this PR is like catnip for me.  I'm swamped, but I definitely want to look more closely at this.  Apologies in advance if I'm delayed in review.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-21_08:19:42-rados-wip-kefu-testing-2021-07-21-1257-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-21_13:52:36-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-21_14:42:35-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/

e2e failures tracked by https://tracker.ceph.com/issues/51728

File: src/common/buffer_pool.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+public:
+    ~BufferPool() {
+      std::lock_guard<std::mutex> l{m_lock};
+      while (!m_buffers.empty()) {
+        auto buf = m_buffers.back();
+        ::operator delete(buf);
+        m_buffers.pop_back();
+      }
+      std::vector<T*> empty;
+      m_buffers.swap(empty);
```
Comment (ronen-fr) [inline src/common/buffer_pool.h:22]: why is this needed? we've already popped all elements out

**Comment by ronen-fr** [comment]
Please don’t consider the following as a criticism. Rather, I’d like to make sure I understand the design decisions that were involved in creating the PR. And I’ll try to explain why I am asking those questions.
Also - please correct me if I am getting some details of the proposed implementation wrong.

1.  one important detail is that the memory pools here are not pre-allocated. Instead - pool memory blocks are “accumulated” from deleted objects. Why was this way chosen?
Note that this decision is, I assume, the reason why ‘object_pools’ has entries for every possible size (with increments of 1). Pool allocators usually group sizes (e.g. by size/NBYTES or by power of 2s).

2. it also means that we are managing pointers to scattered memory blocks - losing for example some debug options (I guess there is no performance penalty here).

3. as we have already templated a specific type allocator on the type - what is gained by having ‘object_pools’? shouldn’t we just drop it? (note that not many types of objects are involved, so there isn’t much to gain from “combining” the memory freed by objects of different types that chanced to have the same size).

4. why was it decided not to use existing pool allocators (e.g. Boost’s)?

5. an alternative design could use the std::pmr interfaces for the specific objects.
Pros: 
ability to play with different strategies (including, for example, “short-sub-object” optimizations, as suggested in our discussions);
existing std collections support;
existing, fully debugged, implementations;
Cons: performance (mostly) (ease-of-use, usually a problem attributed to pmr, is not worse than the proposed PR)

If one accepts the PR design, there are some simple improvements that can be made. For example:
if a BufferPool<T> for a specific T is constructed - why not start with a reasonably-sized vector? after all, if constructing for <T>, it probably means that one of the objects involved is <T>, and we will be allocating many <T>s.

**Comment by orozery** [comment]
Thanks for the feedback everyone!
I'm putting this on hold for now, as new approaches are still being researched.
